File: mcp-server/go/internal/logger/logger.go
// Package logger writes structured log lines to an io.Writer.
//
// The composition root wires this to os.Stderr.  Stdout is the MCP
// transport — a stray write there corrupts the JSON-RPC framing — so
// the logger never accepts a Writer it does not own.
package logger

import (
	"fmt"
	"io"
	"time"
)

// Level is a severity tag printed with every line.
type Level string

const (
	LevelInfo  Level = "INFO"
	LevelWarn  Level = "WARN"
	LevelError Level = "ERROR"
)

// Clock returns the current time.  Injected so tests can assert output.
type Clock func() time.Time

// Stream is the minimal logger the server uses.
type Stream struct {
	w     io.Writer
	clock Clock
}

// New returns a Stream writing to w, using wall-clock time.
func New(w io.Writer) *Stream {
	return &Stream{w: w, clock: time.Now}
}

// NewWithClock is like [New] but with an injectable clock (for tests).
func NewWithClock(w io.Writer, clock Clock) *Stream {
	return &Stream{w: w, clock: clock}
}

// Info writes an INFO line.
func (s *Stream) Info(message string) { s.write(LevelInfo, message) }

// Warn writes a WARN line.
func (s *Stream) Warn(message string) { s.write(LevelWarn, message) }

// Error writes an ERROR line.
func (s *Stream) Error(message string) { s.write(LevelError, message) }

func (s *Stream) write(level Level, message string) {
	_, _ = fmt.Fprintf(
		s.w,
		"%s %s evagene-mcp: %s\n",
		s.clock().UTC().Format(time.RFC3339),
		level,
		message,
	)
}

File: mcp-server/node/src/logger.ts
import type { StructuredLogger } from './server.js';

/**
 * Writes log lines to a stream — **always stderr in production**, since
 * stdout is the MCP transport.  Injecting the stream keeps the module
 * testable and prevents an accidental `console.log` from corrupting the
 * protocol.
 */
export class StreamLogger implements StructuredLogger {
  constructor(private readonly stream: NodeJS.WritableStream) {}

  info(message: string): void {
    this.write('INFO', message);
  }

  warn(message: string): void {
    this.write('WARN', message);
  }

  private write(level: string, message: string): void {
    this.stream.write(`${new Date().toISOString()} ${level} evagene-mcp: ${message}\n`);
  }
}
